src/main/java/com/es/phoneshop/dao/impl/ArrayListProductDao.java

```diff
@@ -105,21 +118,62 @@ public void update(Product product) {
-        save(new Product("nokia3310", "Nokia 3310", new BigDecimal(70), usd, 100, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Nokia/Nokia%203310.jpg"));
-        save(new Product("palmp", "Palm Pixi", new BigDecimal(170), usd, 30, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Palm/Palm%20Pixi.jpg"));
-        save(new Product("simc56", "Siemens C56", new BigDecimal(70), usd, 20, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20C56.jpg"));
-        save(new Product("simc61", "Siemens C61", new BigDecimal(80), usd, 30, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20C61.jpg"));
-        save(new Product("simsxg75", "Siemens SXG75", new BigDecimal(150), usd, 40, "https://raw.githubusercontent.com/andrewosipenko/phoneshop-ext-images/master/manufacturer/Siemens/Siemens%20SXG75.jpg"));
+    @Override
+    public List<Product> findProducts(String name) throws InvalidParameterException {
+        readLock.lock();
+        try {
+            if (name != null) {
+                String words[] = name.split(" ");
+                List<Product> productList = new ArrayList<>();
+                for (String word : words) {
+                    productList.addAll(products.stream()
+                            .filter((product -> product.getDescription().toLowerCase().contains(word.toLowerCase())))
+                            .collect(Collectors.toList()));
+                }
+                productList = productList.stream().distinct().collect(Collectors.toList());
+                productList = productList.stream().sorted(new Comparator<Product>() {
+                    @Override
+                    public int compare(Product o1, Product o2) {
+                        return getNumberOfWordsInjections(words, o2.getDescription())
+                                - getNumberOfWordsInjections(words, o1.getDescription());
+                    }
+                }).collect(Collectors.toList());
+                return productList;
+            }
+            throw new InvalidParameterException();
+        } finally {
+            readLock.unlock();
+        }
+
+    }
+
+    @Override
+    public List<Product> findProducts(String name, String sortField, String sortOrder) throws InvalidParameterException {
+        readLock.lock();
+        try {
+            List<Product> listToSort = findProducts(name);
+            Comparator<Product> priceComparator = Comparator.comparing(product -> product.getPrice());
+            Comparator<Product> descriptionComparator = Comparator.comparing(product -> product.getDescription());
+            Comparator comparator;
+            if (SortOptions.DESCRIPTION.name().equals(sortField.toUpperCase(Locale.ROOT))) {
+                comparator = descriptionComparator;
+            } else {
```

Comment:
Use else if instead of else. Incorrect behavior would happen if sortField is not SortOptions.PRICE